### ISSUE TITLE
Duplicate title mis-numbered

### ISSUE THREAD

**Issue by opoudjis** [metanorma/metanorma-standoc#130]

**Comment by ronaldtse** [comment]
When the section titles are identical, the generated section numbers will be duplicated. 

Specifically, the first instance of the section will share the section number of the last section which has the same title. 

e.g.

<img width="271" alt="Screen Shot 2019-10-05 at 9 01 08 PM" src="https://user-images.githubusercontent.com/11865/66255188-425c0580-e7b3-11e9-922c-f879af3d5b87.png">

<img width="268" alt="Screen Shot 2019-10-05 at 9 02 08 PM" src="https://user-images.githubusercontent.com/11865/66255196-6586b500-e7b3-11e9-9f9d-713abe293dad.png">

In Annex C:

<img width="274" alt="Screen Shot 2019-10-05 at 9 02 19 PM" src="https://user-images.githubusercontent.com/11865/66255197-6c152c80-e7b3-11e9-84e1-120c83f8fc8c.png">

**Comment by opoudjis** [comment]
I cannot replicate in ISO, and there is no obvious reason why this should be happening at all: Asciidoctor assigns distinct sections with the same title distinct anchors, and the headings are being assigned per anchor.

Point me to the document where this is happening. Nothing for me to action until you do.

**Comment by opoudjis** [comment]
This might be csd-digital-addresses, but:

````
 make
mkdir -p documents
make: *** No rule to make target `sources/images/AddressInstanceLifeCycle.png', needed by `sources/images'.  Stop.
````

**Comment by ronaldtse** [comment]
@opoudjis it's fixed, have a look.

**Comment by opoudjis** [comment]
This is an issue with the datamodel plugin, and therefore @PeterTKY should be debugging it.

The issue is that clause numbers are assigned to each clause in isodoc according to their unique anchor ID. So for `<clause id="_conformance" obligation="normative">` in Metanorma XML, clause ID "_conformance" is assigned the clause number "2" in isodoc.

Asciidoctor knows that if two titles are identical, it should assign distinct IDs to them. For:

````
= Doc
  
== Common models

=== Validity

[appendix]
== Annex

=== Validity
````

Asciidoctor assigns ID "_validity" to the first header with that name, and "_validity_2" to the second header with that name. So when isodoc assigns clause numbers to those headers, they have distinct IDs, and are assigned distinct numbers.

The feature/datamodel plugin is not doing this. The second instance of "Validity" in that document is not being inserted by Asciidoctor, so both instances of the header "Validity" have the same id, "_validity".

The solution is for @PeterTKY to implement: whatever is generating those clauses from the UML, should be ensuring its clause IDs are distinct from any clause IDs already in Asciidoc. The simplest way of doing so is to just make the ID a GUID.

**Comment by opoudjis** [comment]
The issue is presumably at:

````
def self.header_to_asciidoc(name, hash, level)
        [
          directive_to_asciidoc(hash),
          "#{"=" * level} #{name}",
        ].compact.join("\n")
      end
````

I see you're authoring raw Asciidoc. This is naively giving it a default ID without checking. So change that, I guess, to:

````
def self.header_to_asciidoc(name, hash, level)
          directive_to_asciidoc(hash),
          "[[_#{UUIDTools::UUID.random_create}]]\n#{"=" * level} #{name}",
        ].compact.join("\n")
      end
````

But do test, and do rspec this. I don't know how this code works.

**Comment by opoudjis** [comment]
Digging deeper, the mistaken assumption is in this code:

````
def self.directive_to_asciidoc(hash)
        directive = (hash || {})["directive"]
        directive ? "[[#{directive}]]" : nil
      end
````

This code is used to explicitly assign the ID "validity" to a UML class generated header "Validity".

The bug here is that the Validity header is being generated in the appendix, and Asciidoctor has already generated the ID "validity" for the header in section 8.

I have no idea how to inspect the document from here for whether IDs have already been assigned to headers, and I don't see any particular need to do so. I am assigning a GUID as the ID of all calls to self.directive_to_asciidoc. If this code wants to assign a different ID explicitly to the clause for crossreferences, then that ID needs to be set in the UML --- or else it needs to create clauses using the Asciidoctor API, which might (*might*) insert disambiguating numbers into the ID automatically.

**Comment by opoudjis** [comment]
This is a patch, and the real solution to this issue is likely #133. But that is not a task I can undertake.

**Comment by opoudjis** [comment]
The lack of any rspec testing for the code in this branch must be remedied before it can be considered for any integration with master branch.

**Comment by opoudjis** [comment]
Oh, it's worse than I thought. The code I thought was responsible for creating the IDs isn't: this code *is* in fact using the API. And it's creating new sections, while ignoring what is already in the document:

````
def models_to_sections(block, view_hash)
        Asciidoctor::DataModel::AsciidocAdaptor.for_each("classes", view_hash) do |class_name, class_hash|
          class_fidelity = class_hash["fidelity"] || {}
          next if class_fidelity["skipSection"]

          section = create_section block, class_name, {}
          ....
        end
````

The command `section = create_section block, class_name, {}` is naively AND INCORRECTLY assigning the ID "_Validate", because it has no idea what is in the rest of the document. That is your problem right there.

The full horror here is block was generated as part of the current asciidoctor section, so this command *should* know what is in the rest of the document: parent.document is in fact appealed to at the start of the macro, and block is generated under parent. So this is likely an Asciidoctor bug. (Likely, because it's not like Asciidoctor code is exhaustively documented.)

Given this likely bug, I simply cannot trust `create_section` to automatically assign IDs to sections, even if it is given information about the  document it is to be included in. I am overriding the automatic assignment of an ID to the section with a GUID, as intended.
 
If the code I was looking at in asciidoc_adaptor.rb is no longer being used, it needs to be removed.

**Comment by PeterTKY** [comment]
> The lack of any rspec testing for the code in this branch must be remedied before it can be considered for any integration with master branch.

Definitely! And that's why we haven't made it into the master branch. The existence of this thing was very adhoc. It was simply copied from the code written to handle a requirement to specify the definitions of attributes in every model for the draft I was writing.

Will have to clean up the code and complete the rspec before merging it to the master branch.

**Comment by opoudjis** [comment]
No, you're doing the right thing, @PeterTKY; I got confused because of the old code. This is an Asciidoctor bug, that it does not generate the IDs correctly, respecting IDs already assigned in the document.
